### PR TITLE
macro_rules: Preserve all metavariable spans in a global side table

### PR DESCRIPTION
**Pull request by petrochenkov**
This PR preserves spans of `tt` metavariables used to pass tokens to declarative macros.
Such metavariable spans can then be used in span combination operations like `Span::to` to improve all kinds of diagnostics.

Spans of non-`tt` metavariables are currently kept in nonterminal tokens, but the long term plan is remove all nonterminal tokens from rustc parser and rely on the proc macro model with invisible delimiters (#114647, #67062).
In particular, `NtIdent` nonterminal (corresponding to `ident` metavariables) becomes easy to remove when this PR lands (#119412 does it).

The metavariable spans are kept in a global side table keyed by `Span`s of original tokens.
The alternative to the side table is keeping them in `SpanData` instead, but the performance regressions would be large because any spans from tokens passed to declarative macros would stop being inline and would work through span interner instead, and the penalty would be paid even if we never use the metavar span for the given original span.
(But also see the comment on `fn maybe_use_metavar_location` describing the map collision issues with the side table approach.)

There are also other alternatives - keeping the metavar span in `Token` or `TokenTree`, but associating it with `Span` itsel is the most natural choice because metavar spans are used in span combining operations, and those operations are not necessarily tied to tokens.

### CODE REVIEW

**Comment by petrochenkov** [comment]
Okay, let's try a light version - the new table insertion logic is disabled for undelimited sequences `$($tt)*`.
@bors try @rust-timer queue

**Comment by petrochenkov** [comment]
While I'm on holidays I'm just going to run rust-perf in the background, without benchmarking anything myself.
@bors try @rust-timer queue

**Comment by petrochenkov** [comment]
It's `marker.visit_span(&mut sp);` that was causing the regressions.
Very good.

**Comment by petrochenkov** [comment]
There's one relatively significant design decision implicitly made by this PR.
It's about spans of AST nodes.

Consider this code:
```rust
macro_rules! pass {
    ($tt:tt) => { $tt }
}

fn main() {
    let b = pass!(a);
}
```

`a` here is 1) an identifier, 2) a path, and 3) an expression.
Each of these entities has its own span in AST and further IRs.

There are two main possibilities for assigning spans here.
- 1. `expr_span = path_span = ident_span = span_of(a)`
- 2. `ident_span = span_of(a)` and `expr_span = path_span = span_of($tt)` 

The first alternative follows naturally from span combining logic in this PR.
Parser typically uses `span_lo.to(span_hi)` for assigning AST node spans, and `span_lo == span_hi == span_of(a)` in case of single-identifier paths, expressions, types, patterns, etc, so the result is `span_of(a)` for everything.

The second alternative is currently used for `ident` metavars aka `NtIdent` (see #119412 for the effect of the new span combining logic on it).

I experimented with both alternatives, and clearly none of them is perfect.
But the first alternative seems preferable overall, it always results in most precise user-provided spans - this is especially visible from the diffs for the `mir!` macro in #119412.

What are the drawbacks of the first alternative?
You cannot use `expr.span.shrink_to_lo()` or `expr.span.shrink_to_hi()` as insertion points for suggestions in general case.
If `a` is passed to a macro as `ident`, then a suggestion to e.g. add a reference to it `&a` will not compile - see test `tests/ui/borrowck/move-error-snippets.stderr` in #119412, for example.

Giving suggestions to insert code inside macro bodies, or near code passed to macros as tokens seems generally questionable though.
If you really want to make such suggestions, you need to look at the code around, like `Span::with_neighbor` does, but even that is not a 100% guarantee.
E.g. in
```rust
macro_rules! mac {
    ($a:tt $b:tt) => {}
}

fn main() {
    let b = mac!(a b);
}
```
if you want to suggest inserting `+` between `a` and `b`, then `mac!(a + b)` seems to be the right answer, because `a` and `b` have matching contexts, but it still won't compile because `mac` takes 2 tokens and not 3.

**Comment by petrochenkov** [comment]
The perf regressions are in acceptable range, and https://github.com/rust-lang/rust/pull/119693 improves the situation further.
@rustbot ready

**Comment by compiler-errors** [comment]
> I experimented with both alternatives, and clearly none of them is perfect.

Yeah. It looks like the diagnostic regressions + improvements generally cancel out here... In general, we should probably be a bit more careful about making suggestions around nonterminals, so I wonder if we should just fix these in diagnostics as we encounter them.

> If you really want to make such suggestions, you need to look at the code around, like Span::with_neighbor does, but even that is not a 100% guarantee.

Yeah, I agree.

**Review by compiler-errors — APPROVED** [review]
pr looks fine to me, but maybe someone else can take a look because of the many *interesting* changes to diagnostics.

I don't disagree that we're making things more consistent here, and perhaps the diagnostics regressions are just a consequence of the compiler having tailored itself very closely to the old behavior of nt spans.

**Comment by compiler-errors** [comment]
r? cjgillot

**Comment by petrochenkov** [comment]
@cjgillot any chance to review or reassign this?

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -100,6 +101,7 @@ mod tests;
 pub struct SessionGlobals {
     symbol_interner: symbol::Interner,
     span_interner: Lock<span_encoding::SpanInterner>,
+    metavar_spans: Lock<FxHashMap<Span, Span>>,
```
Comment (cjgillot) [inline compiler/rustc_span/src/lib.rs:104]: This needs a comment: which span is the key, and which is the value.

**Review by cjgillot — COMMENTED** [review]
Sorry I took so long.

I'm trying to get a grasp of what's happening here. IIUC, the changes to `rustc_expand` are bookkeeping, and the changes to `rustc_span` are meant to use the spans.

When walking contexts from a span with `find_ancestor_{inside,in_same_ctxt}`, do metavariable spans appear anywhere? Should they? Or are the users of that method wrong?

**Comment by petrochenkov** [comment]
>IIUC, the changes to `rustc_expand` are bookkeeping, and the changes to `rustc_span` are meant to use the spans.

Exactly.

>When walking contexts from a span with `find_ancestor_{inside,in_same_ctxt}`, do metavariable spans appear anywhere?

No, the metavariable spans are only used in span combining operations like `span_lo.to(span_hi)`.

The argument spans passed to `find_ancestor_*` functions may change after this PR, if they were constructed using span combining (which is likely), but otherwise `find_ancestor_*` functions are not especially related to this change. Why did you ask about them specifically.

**Comment by cjgillot** [comment]
I've been using those `find_ancestor` for suggestions that require span manipulation. Should they be migrated to a more robust method?

**Comment by petrochenkov** [comment]
> I've been using those `find_ancestor` for suggestions that require span manipulation. Should they be migrated to a more robust method?

`find_ancestor_in*` methods should continue working exactly because they takes some neighboring code into account (as mentioned in https://github.com/rust-lang/rust/pull/119673#issuecomment-1880264729 above).
What will stop working in some cases is naive insertion of suggesions into locations like `expr.span.shrink_to_lo()` or `expr.span.shrink_to_hi()`.

**Comment by cjgillot** [comment]
Thanks.
r=me after rebase

**Comment by nnethercote** [comment]
Some icount regressions here. Was the idea that the perf improvements in #119693 justified the regressions here?

**Comment by petrochenkov** [comment]
If regressions in the final run are smaller than in the previous run (https://github.com/rust-lang/rust/pull/119673#issuecomment-1880097503), then https://github.com/rust-lang/rust/pull/119693 is responsible for that.

In general, with this PR we are doing more work (tracking more spans), so some regressions are expected.
Also this PR enables removal of `NtIdent` (https://github.com/rust-lang/rust/pull/119412), which brings improvements.
(I also have a `span.ctxt` optimization in the pipeline that should help here, but won't be able to work on it earlier than next week.)

**Comment by pnkfelix** [comment]
* Instruction count regressions flagged against serde_derive, diesel, serde, html5ever, syn
* Most of these were anticipated in the final perf runs against the try builds.
* Marking as triaged under assumption that the reviewers deemed this change to be worth the predicted cost.

@rustbot label: +perf-regression-triaged

**Pull request merged**